File: Python/109.Convert Sorted List to Binary Search Tree.py
#根据排序链表生成一个平衡二叉排序树
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def sortedListToBST(self, head):
        nums , p = [] , head
        while p!=None:
            nums.append(p.val)
            p = p.next
        return self.toBST(nums,0,len(nums)-1)

    def toBST(self,nums,l,r):
        if l>r:
            return None
        mid = (l+r)//2
        node = TreeNode(nums[mid])
        node.left = self.toBST(nums,l,mid-1)
        node.right = self.toBST(nums,mid+1,r)
        return node

class Ssolution(object):
    def inorderTraversal(self, root):
        self.result = []
        self.inorder(root)
        return self.result

    def preorderTraversal(self, root):
        self.result = []
        self.preorder(root)
        return self.result

    def inorder(self,p):
        if p!=None:
            self.inorder(p.left) if p.left!=None else None
            self.result.append(p.val)
            self.inorder(p.right) if p.right!=None else None

    def preorder(self,p):
        if p!=None:
            self.result.append(p.val)
            self.preorder(p.left) if p.left!=None else None
            self.preorder(p.right) if p.right!=None else None

l1 = ListNode(1)
l2 = ListNode(2)
l3 = ListNode(3)
l4 = ListNode(5)
l5 = ListNode(8)
l6 = ListNode(9)
l7 = ListNode(15)
l8 = ListNode(30)
l1.next = l2
l2.next = l3
l3.next = l4
l4.next = l5
l5.next = l6
l6.next = l7
l7.next = l8



ss = Ssolution()
s = Solution()
root = s.sortedListToBST(l1)
print(ss.inorderTraversal(root))
print(ss.preorderTraversal(root))

File: C++/Linked List/109.Convert Sorted List to Binary Search Tree.cpp
/*
把有序链表转换为平衡二叉搜索树。
*/
/*
用快慢指针，找到中间，建立结点，递归左右。
 */

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	TreeNode* sortedListToBST(ListNode* head) {
		if (!head)	return NULL;
		if (!head->next)
		{
			TreeNode *root = new TreeNode(head->val);
			return root;
		}
		ListNode *fast = head->next->next,*preslow = head;
		while (fast && fast->next)
		{
			preslow = preslow->next;
			fast = fast->next->next;
		}
		TreeNode *root = new TreeNode(preslow->next->val);
		root->right = sortedListToBST(slow->next->next);
		preslow->next = NULL;
		root->left = sortedListToBST(head);
		return root;
    }
};
int main()
{
	Solution s;
	ListNode head(1);
	ListNode t1(2);
	ListNode t2(3);
	ListNode t3(4);
	ListNode t4(5);
	ListNode t5(6);
	head.next = &t1;
	t1.next = &t2;
	t2.next = &t3;
	t3.next = &t4;
	t4.next = &t5;
 	TreeNode* h = s.sortedListToBST(&head);
}

